File: sunnad-ios/sunnad-ios/Features/Profile/ViewModels/InsightsViewModel.swift
import Combine
import Foundation

struct InsightPoint: Identifiable, Equatable {
    let date: Date
    let completed: Double
    let due: Double
    let missedHabitIDs: [UUID]

    var id: Date { date }
}

struct HabitDayState: Identifiable, Equatable {
    let date: Date
    let scheduled: Bool
    let completed: Bool

    var id: Date { date }
}

struct HabitPerformance: Identifiable, Equatable {
    let id: UUID
    let title: String
    let percentage: Int
    let days: [HabitDayState]
}

struct CategoryPerformance: Identifiable, Equatable {
    let key: String
    let title: String
    let percentage: Int
    let completed: Int
    let total: Int

    var id: String { key }
}

@MainActor
final class InsightsViewModel: ObservableObject {
    @Published private(set) var points: [InsightPoint] = []
    @Published private(set) var habitPerformance: [HabitPerformance] = []
    @Published private(set) var categoryPerformance: [CategoryPerformance] = []
    @Published private(set) var isLoading = false
    @Published private(set) var errorMessage: String?

    private let habitsRepository: HabitsRepository
    private let completionsRepository: CompletionsRepository
    private let logger: AnalyticsLogging
    private let calendar: Calendar
    private let timeZone: TimeZone
    private let now: () -> Date
    private let totalDays = 40

    private var habitsByID: [UUID: Habit] = [:]
    private var habitTitleByID: [UUID: String] = [:]
    private var completionByHabitAndDay: [UUID: [Date: HabitCompletion]] = [:]

    init(
        habitsRepository: HabitsRepository,
        completionsRepository: CompletionsRepository,
        logger: AnalyticsLogging,
        calendar: Calendar = .current,
        timeZone: TimeZone = .current,
        now: @escaping () -> Date = Date.init
    ) {
        self.habitsRepository = habitsRepository
        self.completionsRepository = completionsRepository
        self.logger = logger
        self.calendar = calendar
        self.timeZone = timeZone
        self.now = now
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            let today = now()
            let habits = try await habitsRepository.fetchHabits(includeArchived: false)
            habitsByID = Dictionary(uniqueKeysWithValues: habits.map { ($0.id, $0) })
            habitTitleByID = Dictionary(uniqueKeysWithValues: habits.map { ($0.id, $0.name) })

            var completionLookup: [UUID: [Date: HabitCompletion]] = [:]
            var performances: [HabitPerformance] = []
            var categoryAggregates: [String: (title: String, completedDays: Int, totalDays: Int)] = [:]
            var categoryIdentityByHabitID: [UUID: (key: String, title: String)] = [:]

            let days = timelineDays(asOf: today)
            let daySet = Set(days.map { startOfDay(for: $0) })

            for habit in habits {
                let history = try await completionsRepository.fetchCompletions(for: habit.id)
                let byDay = Dictionary(uniqueKeysWithValues: history.map { (startOfDay(for: $0.dayDate), $0) })
                completionLookup[habit.id] = byDay

                let dayStates = days.map { day in
                    let scheduled = habit.isDue(on: day, calendar: calendar, timeZone: timeZone)
                    let completion = byDay[startOfDay(for: day)]
                    let completed = scheduled && completion?.isCompleted(for: habit) == true
                    return HabitDayState(date: day, scheduled: scheduled, completed: completed)
                }

                let scheduledCount = dayStates.filter(\.scheduled).count
                let completedCount = dayStates.filter(\.completed).count
                let percentage = scheduledCount == 0 ? 0 : Int((Double(completedCount) / Double(scheduledCount) * 100).rounded())

                let customLabel = habit.categoryCustom?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
                let categoryKey: String
                let categoryTitle: String
                if let customLabel {
                    categoryKey = "custom:\(customLabel.lowercased())"
                    categoryTitle = customLabel
                } else {
                    categoryKey = "preset:\(habit.category.rawValue)"
                    categoryTitle = L10n.t("category.\(habit.category.rawValue)")
                }
                categoryIdentityByHabitID[habit.id] = (key: categoryKey, title: categoryTitle)

                performances.append(
                    HabitPerformance(
                        id: habit.id,
                        title: habit.name,
                        percentage: percentage,
                        days: dayStates
                    )
                )
            }

            completionByHabitAndDay = completionLookup
            habitPerformance = performances.sorted { lhs, rhs in
                if lhs.percentage == rhs.percentage {
                    return lhs.title.localizedCompare(rhs.title) == .orderedAscending
                }
                return lhs.percentage > rhs.percentage
            }
            points = days.map { day in
                let scheduledHabits = habits.filter { $0.isDue(on: day, calendar: calendar, timeZone: timeZone) }
                var completedCount = 0
                var missedHabitIDs: [UUID] = []
                var categoryDayRollup: [String: (title: String, due: Int, completed: Int)] = [:]

                for habit in scheduledHabits {
                    let completion = completionLookup[habit.id]?[startOfDay(for: day)]
                    let isCompleted = completion?.isCompleted(for: habit) == true
                    if isCompleted {
                        completedCount += 1
                    } else {
                        missedHabitIDs.append(habit.id)
                    }

                    if let identity = categoryIdentityByHabitID[habit.id] {
                        var bucket = categoryDayRollup[identity.key] ?? (title: identity.title, due: 0, completed: 0)
                        bucket.due += 1
                        if isCompleted {
                            bucket.completed += 1
                        }
                        categoryDayRollup[identity.key] = bucket
                    }
                }

                for (key, bucket) in categoryDayRollup {
                    var aggregate = categoryAggregates[key] ?? (title: bucket.title, completedDays: 0, totalDays: 0)
                    aggregate.totalDays += 1
                    if bucket.due > 0 && bucket.due == bucket.completed {
                        aggregate.completedDays += 1
                    }
                    categoryAggregates[key] = aggregate
                }

                let dueCount = scheduledHabits.count
                let dueValue = max(dueCount, completedCount)

                return InsightPoint(
                    date: day,
                    completed: Double(completedCount),
                    due: Double(dueValue),
                    missedHabitIDs: missedHabitIDs
                )
            }

            categoryPerformance = categoryAggregates.map { key, aggregate in
                let percentage = aggregate.totalDays == 0
                    ? 0
                    : Int((Double(aggregate.completedDays) / Double(aggregate.totalDays) * 100).rounded())
                return CategoryPerformance(
                    key: key,
                    title: aggregate.title,
                    percentage: percentage,
                    completed: aggregate.completedDays,
                    total: min(aggregate.totalDays, totalDays)
                )
            }.sorted { lhs, rhs in
                if lhs.percentage == rhs.percentage {
                    return lhs.title.localizedCompare(rhs.title) == .orderedAscending
                }
                return lhs.percentage > rhs.percentage
            }

            // Keep only data inside the active 40-day window to reduce memory churn.
            completionByHabitAndDay = completionByHabitAndDay.mapValues { entries in
                entries.filter { daySet.contains($0.key) }
            }
        } catch {
            errorMessage = error.localizedDescription
            logger.log(.storageFailure, metadata: ["scope": "insights_load", "error": error.localizedDescription])
            points = []
            habitPerformance = []
            categoryPerformance = []
        }
    }

    func missedHabitTitles(for point: InsightPoint) -> [String] {
        point.missedHabitIDs.compactMap { habitTitleByID[$0] }
    }

    private func timelineDays(asOf today: Date) -> [Date] {
        let todayStart = startOfDay(for: today)
        return (0..<totalDays).compactMap { index in
            calendar.date(byAdding: .day, value: -(totalDays - 1 - index), to: todayStart)
        }
    }

    private func startOfDay(for date: Date) -> Date {
        var calendar = calendar
        calendar.timeZone = timeZone
        return calendar.startOfDay(for: date)
    }
}

private extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}

File: sunnad-android/app/src/main/java/com/arystan/almasuly/sunnadandroid/features/insights/InsightsViewModel.kt
package com.arystan.almasuly.sunnadandroid.features.insights

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arystan.almasuly.sunnadandroid.core.model.HabitCategoryValue
import com.arystan.almasuly.sunnadandroid.core.model.Habit
import com.arystan.almasuly.sunnadandroid.domain.repository.CompletionsRepository
import com.arystan.almasuly.sunnadandroid.domain.repository.HabitsRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.util.UUID

data class InsightsPointUi(
    val date: LocalDate,
    val due: Int,
    val completed: Int,
    val missedHabitIds: List<UUID>
)

data class HabitDayUi(
    val date: LocalDate,
    val scheduled: Boolean,
    val completed: Boolean
)

data class HabitPerformanceUi(
    val id: UUID,
    val title: String,
    val percentage: Int,
    val days: List<HabitDayUi>
)

data class CategoryPerformanceUi(
    val key: String,
    val category: HabitCategoryValue?,
    val customLabel: String?,
    val percentage: Int,
    val completed: Int,
    val total: Int
)

data class InsightsUiState(
    val points: List<InsightsPointUi> = emptyList(),
    val habitPerformance: List<HabitPerformanceUi> = emptyList(),
    val categoryPerformance: List<CategoryPerformanceUi> = emptyList(),
    val habitTitlesById: Map<UUID, String> = emptyMap(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null
)

class InsightsViewModel(
    private val habitsRepository: HabitsRepository,
    private val completionsRepository: CompletionsRepository
) : ViewModel() {
    private val _state = MutableStateFlow(InsightsUiState())
    val state: StateFlow<InsightsUiState> = _state.asStateFlow()

    private val totalDays = 40

    fun load() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, errorMessage = null) }
            runCatching {
                val habits = habitsRepository.fetchHabits(includeArchived = false)
                val timeline = timelineDays(LocalDate.now())
                val completionByHabitId = mutableMapOf<UUID, Map<LocalDate, Int>>()
                data class CategoryAggregate(
                    val category: HabitCategoryValue?,
                    val customLabel: String?,
                    var completedDays: Int = 0,
                    var totalDays: Int = 0
                )
                val categoryAggregates = linkedMapOf<String, CategoryAggregate>()

                habits.forEach { habit ->
                    val completionMap = completionsRepository
                        .fetchCompletions(habit.id)
                        .associate { it.dayDate to it.value }
                    completionByHabitId[habit.id] = completionMap
                }

                val points = timeline.map { day ->
                    val dueHabits = habits.filter { it.isDue(day) }
                    var completedCount = 0
                    val missedIds = mutableListOf<UUID>()
                    val categoryDayStatus = linkedMapOf<String, Pair<Int, Int>>()

                    dueHabits.forEach { habit ->
                        val value = completionByHabitId[habit.id]?.get(day) ?: 0
                        val completed = if (habit.isDhikr) value >= habit.normalizedTargetCount else value >= 1
                        if (completed) completedCount += 1 else missedIds += habit.id

                        val identity = categoryIdentity(habit)
                        val aggregate = categoryAggregates.getOrPut(identity.key) {
                            CategoryAggregate(
                                category = identity.category,
                                customLabel = identity.customLabel
                            )
                        }
                        val dayPair = categoryDayStatus[identity.key] ?: (0 to 0)
                        categoryDayStatus[identity.key] = if (completed) {
                            (dayPair.first + 1) to (dayPair.second + 1)
                        } else {
                            (dayPair.first + 1) to dayPair.second
                        }
                        // Keep aggregate's metadata aligned with the latest observed habit label/category.
                        if (aggregate.customLabel == null && identity.customLabel != null) {
                            categoryAggregates[identity.key] = aggregate.copy(customLabel = identity.customLabel)
                        }
                    }

                    categoryDayStatus.forEach { (key, pair) ->
                        val aggregate = categoryAggregates[key] ?: return@forEach
                        aggregate.totalDays += 1
                        if (pair.first > 0 && pair.first == pair.second) {
                            aggregate.completedDays += 1
                        }
                    }

                    InsightsPointUi(
                        date = day,
                        due = dueHabits.size,
                        completed = completedCount,
                        missedHabitIds = missedIds
                    )
                }

                val performance = habits.map { habit ->
                    val dayStates = timeline.map { day ->
                        val scheduled = habit.isDue(day)
                        val value = completionByHabitId[habit.id]?.get(day) ?: 0
                        val completed = if (habit.isDhikr) value >= habit.normalizedTargetCount else value >= 1
                        HabitDayUi(day, scheduled, scheduled && completed)
                    }

                    val scheduledCount = dayStates.count { it.scheduled }
                    val completedCount = dayStates.count { it.completed }
                    val percentage = if (scheduledCount == 0) 0 else {
                        ((completedCount.toDouble() / scheduledCount.toDouble()) * 100.0).toInt()
                    }

                    HabitPerformanceUi(
                        id = habit.id,
                        title = habit.name,
                        percentage = percentage,
                        days = dayStates
                    )
                }.sortedWith(compareByDescending<HabitPerformanceUi> { it.percentage }.thenBy { it.title })

                val categoryPerformance = categoryAggregates.map { (key, aggregate) ->
                    val percentage = if (aggregate.totalDays == 0) 0 else {
                        ((aggregate.completedDays.toDouble() / aggregate.totalDays.toDouble()) * 100.0).toInt()
                    }
                    CategoryPerformanceUi(
                        key = key,
                        category = aggregate.category,
                        customLabel = aggregate.customLabel,
                        percentage = percentage,
                        completed = aggregate.completedDays,
                        total = aggregate.totalDays.coerceAtMost(totalDays)
                    )
                }.sortedWith(
                    compareByDescending<CategoryPerformanceUi> { it.percentage }.thenBy {
                        it.customLabel ?: it.category?.name ?: it.key
                    }
                )

                val titles = habits.associate { it.id to it.name }
                Quad(points, performance, categoryPerformance, titles)
            }.onSuccess { (points, performance, categoryPerformance, titles) ->
                _state.update {
                    it.copy(
                        points = points,
                        habitPerformance = performance,
                        categoryPerformance = categoryPerformance,
                        habitTitlesById = titles,
                        isLoading = false,
                        errorMessage = null
                    )
                }
            }.onFailure { error ->
                _state.update {
                    it.copy(
                        isLoading = false,
                        categoryPerformance = emptyList(),
                        errorMessage = error.localizedMessage ?: "Insights load failed"
                    )
                }
            }
        }
    }

    private fun timelineDays(today: LocalDate): List<LocalDate> {
        return (0 until totalDays).map { index ->
            today.minusDays((totalDays - 1L) - index.toLong())
        }
    }

    fun missedHabitTitles(point: InsightsPointUi): List<String> {
        return point.missedHabitIds.mapNotNull { _state.value.habitTitlesById[it] }
    }
}

private data class CategoryIdentity(
    val key: String,
    val category: HabitCategoryValue?,
    val customLabel: String?
)

private fun categoryIdentity(habit: Habit): CategoryIdentity {
    val customLabel = habit.categoryCustom?.trim().takeUnless { it.isNullOrEmpty() }
    return if (customLabel != null) {
        CategoryIdentity(
            key = "custom:${customLabel.lowercase()}",
            category = null,
            customLabel = customLabel
        )
    } else {
        CategoryIdentity(
            key = "preset:${habit.category.name.lowercase()}",
            category = habit.category,
            customLabel = null
        )
    }
}

private data class Quad<A, B, C, D>(
    val first: A,
    val second: B,
    val third: C,
    val fourth: D
)
